File: OCalm/src/Jeu_min_max.ml

(* # Importations *)
Random.self_init()

(* # Données *)
let chemin_dossier = Filename.dirname Sys.argv.(0)
let fichier_options = Filename.concat chemin_dossier "minmax_options.txt"

(* # Fonctions utilitaires *)
	(**	Convertit la saisie utilisateur en entier sécurisé
		Valide que l'entier est compris entre min et max
		@message message affiché
		@min borne minimale
		@max borne maximale	*)
	let lire_entier message min max =
		let rec boucle() =
			Printf.printf "%s%!" message;
			let saisie = read_line() in
			match int_of_string_opt saisie with
			| Some valeur when valeur >= min && valeur <= max -> valeur
			| _ ->
				Printf.printf "\nVeuillez entrer un entier entre %d et %d.\n%!" min max;
				boucle()
		in
		boucle()

	(**	Charge les options sauvegardées ou valeurs par défaut
		Retourne (min, max, tours)	*)
	let charger_options() =
		try
			let canal = open_in fichier_options in
			let ligne = input_line canal in
			close_in canal;
			match List.map int_of_string (String.split_on_char ',' ligne) with
			| [min; max; tours] -> (min, max, tours)
			| _ -> (1, 100, 5)
		with Sys_error _ -> (1, 100, 5)

	(**	Sauvegarde les options dans le fichier dédié
		@min borne minimale
		@max borne maximale
		@tours nombre de tentatives	*)
	let sauvegarder_options min max tours =
		let canal = open_out fichier_options in
		Printf.fprintf canal "%d,%d,%d" min max tours;
		close_out canal

(* # Fonctions principales *)
	(**	Affiche le menu principal	*)
	let afficher_menu_principal() =
		Printf.printf "\n#### Menu ####\n1 : Jouer\n2 : Options\n3 : Quitter\n%!"

	(**	Affiche le menu des options
		@min borne minimale
		@max borne maximale
		@tours nombre de tentatives	*)
	let afficher_menu_options min max tours =
		Printf.printf
			"\n#### Menu options ####\n1 : Choisir les limites (actuellement : %d - %d)\n2 : Nombre de tours max (actuellement : %d)\n%!"
			min max tours

	(**	Lance une partie de devinette
		@min borne minimale
		@max borne maximale
		@tours nombre de tentatives	*)
	let jouer_partie min max tours =
		let cible = Random.int (max - min + 1) + min in
		Printf.printf "\nTrouvez entre %d et %d en %d tours\n%!" min max tours;
		for iterator_numero = 1 to tours do
			let choix = lire_entier (Printf.sprintf "Tour %d : " iterator_numero) min max in
			if choix = cible then (Printf.printf "\nGagné !\n%!"; raise Exit);
			Printf.printf "%s\n%!" (if choix > cible then "-" else "+")
		done;
		Printf.printf "\nPerdu ! La réponse était %d.\n%!" cible

	(**	Gère la modification des options et leur sauvegarde
		@min borne minimale
		@max borne maximale
		@tours nombre de tentatives	*)
	let gerer_options min max tours =
		afficher_menu_options min max tours;
		match lire_entier "? = " 1 2 with
		| 1 ->
			let nouveau_min = lire_entier "\nMin = " 1 max in
			let nouveau_max = lire_entier "\nMax = " nouveau_min 10_000 in
			sauvegarder_options nouveau_min nouveau_max tours;
			(nouveau_min, nouveau_max, tours)
		| 2 ->
			let nouveau_tours = lire_entier "\nTours max = " 1 100 in
			sauvegarder_options min max nouveau_tours;
			(min, max, nouveau_tours)
		| _ -> (min, max, tours)

(* # Main *)
let rec main() =
	let min, max, tours = charger_options() in
	let rec boucle min_bound max_bound tours_bound =
		afficher_menu_principal();
		match lire_entier "? = " 1 3 with
		| 1 -> (try jouer_partie min_bound max_bound tours_bound with Exit -> ()); boucle min_bound max_bound tours_bound
		| 2 ->
			let nouveau_min, nouveau_max, nouveau_tours = gerer_options min_bound max_bound tours_bound in
			boucle nouveau_min nouveau_max nouveau_tours
		| 3 -> ()
		| _ -> boucle min_bound max_bound tours_bound
	in
	boucle min max tours

(* # Lancement du programme *)
let () = main()

File: Fortran/src/Jeu_min_max.f90
! # Importations
module minmax_mod
	implicit none
	private
	public	:: main

! # Données
	character(len=*), parameter :: fichier_options = 'data/minmax_options.txt'

! # Fonctions utilitaires
	contains

	! ## Lecture d'un entier utilisateur sécurisé
	!**
	! Demande un entier borné jusqu’à obtention d’une entrée valide
	! @message	texte d’invite
	! @borne_min	valeur minimale
	! @borne_max	valeur maximale
	! @resultat	entier valide saisi
	!**
	subroutine lire_entier(message, borne_min, borne_max, resultat)
		implicit none
		character(len=*), intent(in)	:: message
		integer, intent(in)				:: borne_min, borne_max
		integer, intent(out)			:: resultat
		character(len=32)				:: tampon
		integer							:: valeur, erreur
		do
			write(*,'(A)',advance='no') message
			read(*,'(A)') tampon
			read(tampon,*,iostat=erreur) valeur
			if(erreur==0 .and. valeur>=borne_min .and. valeur<=borne_max)then
				resultat = valeur
				exit
			endif
			write(*,'(A)') ''
			write(*,'(A,I0,A,I0,A)') 'Veuillez entrer un entier entre ',borne_min,' et ',borne_max,'.'
		end do
	end subroutine lire_entier

	! ## Conversion entier → chaîne
	function int_to_str(valeur) result(chaine)
		implicit none
		integer, intent(in)		:: valeur
		character(len=32)		:: chaine
		write(chaine,'(I0)') valeur
		chaine = trim(adjustl(chaine))
	end function int_to_str

	! ## Chargement des options sauvegardées
	subroutine charger_options(borne_min, borne_max, limites_tours)
		implicit none
		integer, intent(out)	:: borne_min, borne_max, limites_tours
		integer					:: unite, statut
		logical					:: existe
		inquire(file=fichier_options,exist=existe)
		if(existe)then
			open(newunit=unite,file=fichier_options,status='old',action='read')
			read(unite,*,iostat=statut) borne_min, borne_max, limites_tours
			close(unite)
			if(statut==0) return
		endif
		borne_min = 1;	borne_max = 100;	limites_tours = 5
	end subroutine charger_options

	! ## Sauvegarde des options
	subroutine sauvegarder_options(borne_min, borne_max, limites_tours)
		implicit none
		integer, intent(in)	:: borne_min, borne_max, limites_tours
		integer				:: unite
		open(newunit=unite,file=fichier_options,status='replace',action='write')
		write(unite,'(I0,",",I0,",",I0)') borne_min, borne_max, limites_tours
		close(unite)
	end subroutine sauvegarder_options

! # Fonctions principales
	! Affiche le menu principal
	subroutine afficher_menu_principal()
		implicit none
		write(*,'(A)') ''
		write(*,'(A)') '#### Menu ####'
		write(*,'(A)') '1 : Jouer'
		write(*,'(A)') '2 : Options'
		write(*,'(A)') '3 : Quitter'
	end subroutine afficher_menu_principal

	! Affiche le menu d’options
	subroutine afficher_menu_options(borne_min, borne_max, limites_tours)
		implicit none
		integer, intent(in)	:: borne_min, borne_max, limites_tours
		write(*,'(A)') ''
		write(*,'(A)') '#### Menu options ####'
		write(*,'(A,I0,A,I0,A)') '1 : Choisir les limites (actuellement : ',borne_min,' - ',borne_max,')'
		write(*,'(A,I0,A)') '2 : Nombre de tours max (actuellement : ',limites_tours,')'
	end subroutine afficher_menu_options

	! Joue une partie, indique si plus grand / plus petit
	subroutine jouer_partie(borne_min, borne_max, limites_tours)
		implicit none
		integer, intent(in)	:: borne_min, borne_max, limites_tours
		integer				:: cible, essai, choix
		real				:: alea
		call random_number(alea);	cible = int(alea*(borne_max-borne_min+1))+borne_min
		write(*,'(A,A,I0,A,I0,A,I0,A)') '', 'Trouvez entre ', borne_min, ' et ', borne_max, ' en ', limites_tours, ' tours'
		do essai = 1, limites_tours
			call lire_entier('Tour '//trim(int_to_str(essai))//' : ', borne_min, borne_max, choix)
			if(choix==cible)then
				write(*,'(A)') '';	write(*,'(A)') 'Gagné !';	return
			elseif(choix>cible)then
				write(*,'(A)') '-'
			else
				write(*,'(A)') '+'
			endif
		end do
		write(*,'(A,A,I0,A)') '', 'Perdu ! La réponse était ', cible, '.'
	end subroutine jouer_partie

	! Gère les options et sauvegarde si modifiées
	subroutine gerer_options(borne_min, borne_max, limites_tours)
		implicit none
		integer, intent(inout)	:: borne_min, borne_max, limites_tours
		integer					:: choix_options
		call afficher_menu_options(borne_min, borne_max, limites_tours)
		call lire_entier('? = ',1,2,choix_options)
		if(choix_options==1)then
			call lire_entier('Min = ',1,borne_max,borne_min)
			call lire_entier('Max = ',borne_min,10000,borne_max)
		elseif(choix_options==2)then
			call lire_entier('Tours max = ',1,100,limites_tours)
		endif
		call sauvegarder_options(borne_min,borne_max,limites_tours)
	end subroutine gerer_options

! # Main
	subroutine main()
		implicit none
		integer :: borne_min, borne_max, limites_tours, choix_menu
		call charger_options(borne_min,borne_max,limites_tours)
		do
			call afficher_menu_principal()
			call lire_entier('? = ',1,3,choix_menu)
			select case(choix_menu)
				case(1);	call jouer_partie(borne_min,borne_max,limites_tours)
				case(2);	call gerer_options(borne_min,borne_max,limites_tours)
				case(3);	exit
			end select
		end do
	end subroutine main
end module minmax_mod

! # Lancement du programme
program run
	use minmax_mod
	call main()
end program run
